tools/wavetool.py

```diff
@@ -0,0 +1,146 @@
+#! /usr/bin/python3
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright(c) 2020 Intel Corporation. All rights reserved.
+
+"""
+A Tool to Generate and Manipulate Wave Files
+"""
+
+import os
+import sys
+import argparse
+import numpy as np
+import scipy.signal as signal
+import scipy.io.wavfile as wavefile
+
+eps = 0.0000000001 # used to avoid zero division error
+
+def generate_sinusoids(**p):
+    """
+    Generate sine or cosine wave.
+
+    ``y(t) = A * sin(2 * pi * f * t + phi)``
+
+    ``y(t) = A * cos(2 * pi * f * t + phi)``
+
+    Parameters
+    ----------
+    dict of sinusiod parameters:
+    p['type']: Wave type, ``sine`` or ``cosine``
+    p['amp']: Amplitude
+    p['freq']: Frequency
+    p['phase']: Phase
+    p['chan']: Channels
+    p['sample_rate']: Sample rate
+    p['duration']: Duration
+
+    Returns
+    ----------
+    Real n-D Array with shape ``(duration * sample_rate, channel)``
+    """
+    func = np.sin if p['type'] == 'sine' else np.cos
+    time = np.arange(0, p['duration'], 1.0 / p['sample_rate'], dtype=np.float32)
+    data = p['amp'] * func(2 * np.pi * p['freq'] * time + p['phase'])
+    return np.reshape(np.repeat(data, p['chan']),[len(data), p['chan']])
+
+def generate_wav(**cmd):
+    if cmd['generate'] in ['sine', 'cosine']:
+        wave_param = {
+            'type': cmd['generate'], 'amp': cmd['amp'][0], 'freq': cmd['freq'][0],
+            'phase': cmd['phase'][0], 'chan': cmd['channel'], 'sample_rate': cmd['sample_rate'],
+            'duration': cmd['duration'][0]
+        }
+        wave_data = generate_sinusoids(**wave_param)
+        return wave_data
+
+def save_wave(wave_data, **cmd):
+    wave_path = cmd['output'].strip()
+    if wave_path == '.' or not wave_path.endswith('wav'):
+        wave_path = wave_path + '/tmp.wav'
+    sample_bits = cmd['bits']
+    np_types = {'S8': np.int8, 'S16': np.int16, 'S32': np.int32}
+    if sample_bits in np_types.keys():
+        wave_data = (np.iinfo(np_types[sample_bits]).max * wave_data).astype(np_types[sample_bits])
+    wavefile.write(wave_path, cmd['sample_rate'], wave_data)
+
+def do_wave_compare(**cmd):
+    fs_wav, wave = wavefile.read(cmd['recorded_wave'].strip())
+    fs_ref, ref_wave = wavefile.read(cmd['reference_wave'].strip())
+    if fs_wav != fs_ref:
+        print('Can not compare wave with different sample rate')
+        sys.exit(1)
+    if cmd['compare'] == 'smart_amp':
+        compare_wav_smart_amp(wave, ref_wave, fs_wav)
+
+# remove digital zeros in two sides
+# amplitude below 100 are regarded as zero, this is true for wave go through DAC-ADC
+def trim_wave(wave):
+    wave_mono = wave[:,0]
+    left_idx = 0
+    right_idx = wave_mono.shape[0] - 1
+    while True:
+        if abs(wave_mono[left_idx]) > 100:
+            break
+        left_idx = left_idx + 1
+    while True:
+        if abs(wave_mono[right_idx]) > 100:
```

Comment:
I guess this 100 value should be at least a named constant. Maybe even a script parameter?